go/vt/vtgate/planbuilder/operators/insert.go

```diff
@@ -109,6 +109,16 @@ func createOperatorFromInsert(ctx *plancontext.PlanningContext, ins *sqlparser.I
 
 	vTbl, routing := buildVindexTableForDML(ctx, tableInfo, qt, "insert")
 
+	if tableInfo.GetVindexTable().Type == vindexes.TypeReference {
```

Comment:
The reference table source should be changed inside otherwise the vindex table and routing will be pointing to `reference` table and not the `source`

Reply:
Updated code as per our discussion.